src/gadgets/sk_knowledge/sk_know_gadget.rs

```diff
@@ -1,22 +1,122 @@
-use bulletproofs::r1cs::{ConstraintSystem, Variable};
-use zerocaf::edwards::{AffinePoint, EdwardsPoint};
-use zerocaf::field::FieldElement as Fq;
-use zerocaf::ristretto::RistrettoPoint;
+use crate::{ristretto_point::SonnyRistrettoPointGadget, util::binary_constrain_gadget};
+use bulletproofs::{
+    r1cs::{ConstraintSystem, LinearCombination, Prover, R1CSError, R1CSProof, Variable, Verifier},
+    BulletproofGens, PedersenGens,
+};
+use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
+use merlin::Transcript;
+use zerocaf::{ristretto::RistrettoPoint as SonnyRistrettoPoint, scalar::Scalar as SonnyScalar};
 
 pub fn sk_knowledge_gadget(
-    cs: &mut ConstraintSystem,
-    scalar: &[Variable],
-    pk: &[Variable],    // 4 Variables -> Do Struct.
-    q: &mut [Variable], // 4 Variables -> Do Struct.
-    n: &mut [Variable], // 4 Variables -> Do Struct.
+    cs: &mut dyn ConstraintSystem,
+    basep: SonnyRistrettoPointGadget,
+    pk: SonnyRistrettoPointGadget,
+    mut sk: Vec<Variable>,
 ) {
-    for bit in scalar {
-        // Apply conditional selection to bit.
+    // Generate Identity point without the ristretto constraint
+    let mut Q = SonnyRistrettoPointGadget {
+        X: LinearCombination::from(Scalar::zero()),
+        Y: LinearCombination::from(Scalar::one()),
+        Z: LinearCombination::from(Scalar::one()),
+        T: LinearCombination::from(Scalar::zero()),
+    };
+    // Compute pk'
+    sk.reverse();
+    for var in sk {
+        // Check that var is either `0` or `1`
+        binary_constrain_gadget(cs, var);
+        Q = Q.double(cs);
+        // If bit == 1 -> Q = Q + basep
+        let basep_or_id = basep.conditionally_select(LinearCombination::from(var), cs);
+        Q = Q.add(cs, basep_or_id);
     }
-    // Constraint final result of Q against Pk.
+    // Constraint pk' == pk
+    pk.equals(cs, Q);
 }
-// Poin Addition P1 + P2 = P3
-// Point Doubling 2*P1 = P2
-// Conditionally selection -> 1 | 0 -> Do somethig
 
-// 1 0 0 1 1 1 0 1 ...... 0 1 1 0
+fn sk_knowledge_proof(
+    pc_gens: &PedersenGens,
+    bp_gens: &BulletproofGens,
+    basep: SonnyRistrettoPoint,
+    pk: SonnyRistrettoPoint,
+    sk: &[Scalar],
+) -> Result<(R1CSProof, Vec<CompressedRistretto>), R1CSError> {
+    // Generate transcript
+    let mut transcript = Transcript::new(b"Sk_knowledge");
```

Comment:
Just a very minor issue, but generally we have two strategies for the transcript: either we receive it via parameter, as in common dalek's implementations (e.g. zkp crate), or we create a function to return the bp_gens, transcript and pc_gens (my personal favorite)

Reply:
Not sure I'm following. 

This function is not exposed, it only pretends to be an example of how to implement a `sk_knowledge_proof`.
You should implement it on the way you want on your own crates imo.